### PR TITLE
Add `restore` step

### LINKED ISSUE #148: [8pt] "Restore from backup" playbook
Per replicaset - probably

### PR DESCRIPTION
**Pull request by hackallcode**
Closes #148 

Before this patch, it was possible to backup instances, but it wasn't possible to restore instances from these backups.

Now it's possible to restore instance from the last or selected backup by the `restore` step. Backup should be in TGZ or directory format.

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -8,6 +8,7 @@ cartridge_not_save_cookie_in_app_config: false
 cartridge_remove_temporary_files: false
 cartridge_ignore_split_brain: false
 cartridge_paths_to_keep_on_cleanup: []
+cartridge_paths_to_keep_on_restore: []
```
Comment (dokshina) [inline defaults/main.yml]: `to_keep_before_restore`?

File: defaults/main.yml [hunk 2]
```diff
@@ -204,6 +205,9 @@ cartridge_eval_delay: 5
 cartridge_remote_backups_dir: /opt/tarantool/backups
 cartridge_fetch_backups: false
 cartridge_fetch_backups_dir: backups/
+cartridge_restore_backup_path: null
+cartridge_force_restore: false
+cartridge_ignore_alien_backup: false
```
Comment (dokshina) [inline defaults/main.yml]: `_allow_`

Comment (dokshina) [reply defaults/main.yml]: Maybe we should remove this flag?

File: doc/backups.md [hunk 3]
```diff
@@ -41,7 +44,51 @@ For a stateboard instance only snapshot files and instance configuration file wi
...
+* find TGZ backup or folder with backup ([details here](#find-backup));
+* check for conflicting files ([details here](#check-for-conflicting-files));
+* replace current files with files from backup ([details here](#replace-current-files-with-files-from-backup)).
+
+Examples of restoring from backup, you can see [below](#using-restore-step).
+
+### Find backup
+
+1. You can pass path of backup by `cartridge_restore_backup_path` variable.
+   Just set the value of this variable to full path of TGZ or folder backup.
```
Comment (dokshina) [inline doc/backups.md]: Path to a TGZ archive or a directory with backup can be specified.

File: doc/backups.md [hunk 4]
```diff
@@ -41,7 +44,51 @@ For a stateboard instance only snapshot files and instance configuration file wi
...
+2. If `cartridge_restore_backup_path` variable is omitted, the role
+   will look in the `cartridge_remote_backups_dir` folder
+   (by default `/opt/tarantool/backups`) for the backup
+   whose name is the most recent alphabetically and contains the instance ID.
```
Comment (dokshina) [inline doc/backups.md]: was created earlier?

File: doc/steps.md [hunk 5]
```diff
@@ -571,7 +573,22 @@ Output facts:
 
 ### Step `backup_stop`
 
-Stop started [backup](/doc/backup.md) on the instance.
+Stop started [backup](/doc/backups.md) on the instance.
+
+### Step `restore`
+
+Restore each instance from a [backup](#step-backup) archive.
```
Comment (dokshina) [inline doc/steps.md]: `[Restore](/doc/backups.md)`

File: library/cartridge_restore_instance.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        head, tail = os.path.split(head)
+
+    if head and tail:
+        make_dirs(head, uid, gid, mode)
+
+    os.mkdir(path, mode)
+    os.chown(path, uid, gid)
+
+
+def md5(fd):
```
Comment (dokshina) [inline library/cartridge_restore_instance.py]: `md5_reader`?

File: library/cartridge_restore_instance.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        )
+
+    for path in sorted(os.listdir(remote_backups_dir), reverse=True):
+        full_path = os.path.join(remote_backups_dir, path)
+
+        if not os.path.isfile(full_path):
+            continue
+
+        if instance_id in full_path:
+            helpers.warn("Backup '%s' was selected from folder with backups to restore instance" % full_path)
```
Comment (dokshina) [inline library/cartridge_restore_instance.py]: is used

File: library/cartridge_restore_instance.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+def get_folder_conflicting_files(backup_path, ignore_func, instance_id):
+    conflicting_files = []
+    is_backup_of_correct_instance = False
+
+    for root, _, files in os.walk(backup_path):
+        for src in files:
+            full_src = os.path.join(root, src)
+            full_dst = os.path.join('/', os.path.relpath(full_src, backup_path))
+
+            if instance_id in full_dst:
```
Comment (dokshina) [inline library/cartridge_restore_instance.py]: `instance-1` and `instance-11`...

File: molecule/backup/include/check_restore.yml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+- name: 'Remove vinyl directory to test directories restoring'
+  file:
+    path: '{{ cartridge_vinyl_dir_parent }}'
+    state: 'absent'
+  register: backup_archive_res
+
+- name: 'Call restore from last TGZ'
+  vars:
+    cartridge_restore_backup_path: null
```
Comment (dokshina) [inline molecule/backup/include/check_restore.yml:11]: Please report an issue to Ansible!

File: doc/backups.md [hunk 10]
```diff
@@ -51,21 +51,21 @@ For a stateboard instance only snapshot files and instance configuration file wi
...
-   Just set the value of this variable to full path of TGZ or folder backup.
+1. You can pass path to backup by `cartridge_restore_backup_path` variable.
+   Path to a TGZ archive or a directory with backup can be specified.
 2. If `cartridge_restore_backup_path` variable is omitted, the role
-   will look in the `cartridge_remote_backups_dir` folder
+   will look in the `cartridge_remote_backups_dir` directory
    (by default `/opt/tarantool/backups`) for the backup
-   whose name is the most recent alphabetically and contains the instance ID.
-3. If no file or folder is selected, an error will occur.
+   whose name contains the instance ID and is the latest alphabetically (with the biggest date).
```
Comment (dokshina) [inline doc/backups.md]: latest date?

File: doc/backups.md [hunk 11]
```diff
@@ -51,21 +51,21 @@ For a stateboard instance only snapshot files and instance configuration file wi
...
+3. If no file or directory is selected, an error will occur.
```
Comment (dokshina) [inline doc/backups.md]: What does `selected` mean? Maybe `If not found any file or directory backup that belongs to given instance, ...`?

File: doc/steps.md [hunk 12]
```diff
@@ -571,7 +573,23 @@ Output facts:
...
+Stop started [backup](/doc/backups.md) on the instance.
+
+### Step `restore`
+
+[Restore](/doc/backups.md#how-to-restore-from-saved-backup) each instance from a [backup](#step-backup) archive.
+
+Input variables from config:
+
+- `cartridge_paths_to_keep_before_restore` - list of full paths or relative paths
+  to work/memtx/vinyl/wal directory that should be kept on instance restore
```
Comment (dokshina) [inline doc/steps.md]: Let's use \`\` for `work/memtx/vinyl/wal`.

list of paths that are absolute or relative to ... ?

File: doc/steps.md [hunk 13]
```diff
@@ -571,7 +573,23 @@ Output facts:
...
+  (`.tarantool.cookie` will be kept independently of this variable);
+  it's possible to use bash patterns, e.g. `*.control`.
+- `cartridge_restore_backup_path` - path to the instance backup archive on the remote machine;
+- `cartridge_remote_backups_dir` - directory with backups on the remote;
+- `cartridge_force_restore` - flag indicates that conflicting files should be overwritten;
```
Comment (dokshina) [inline doc/steps.md:590]: BTW, what if somebody wants to restore all files except some special? I mean, not rewrite some files from backup?

File: doc/variables.md [hunk 14]
```diff
@@ -203,5 +204,17 @@ For more details see [scenario documentation](/doc/scenario.md).
...
+  machine;
+- `cartridge_force_restore` (`boolean`, default: `false`): flag indicates that conflicting files
+  should be overwritten;
+- `cartridge_allow_alien_backup` (`boolean`, default: `false`): flag indicates that backup of
+  instance with another name can be used;
+- `cartridge_skip_cleanup_on_restore` (`boolean`, default: `false`): flag indicates that cleanup
+  before restoring should be skipped;
+- `cartridge_paths_to_keep_before_restore` (`list-of-strings`, default: `[]`): list of full paths or
+  relative paths to work/memtx/vinyl/wal directory that should be kept on instance
+  restore (`.tarantool.cookie` will be kept independently of this variable); it's possible to use
```
Comment (dokshina) [inline doc/variables.md]: "Should be kept" means that it wouldn't be removed before restore, or it wouldn't be rewritten by backup?

File: library/cartridge_restore_instance.py [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+
+def md5_buffer(fd):
+    hash_md5 = hashlib.md5()
+    for chunk in iter(lambda: fd.read(4096), b""):
+        hash_md5.update(chunk)
+    fd.close()
+    return hash_md5.hexdigest()
+
+
+def is_path_of_instance(path, instance_id):
```
Comment (dokshina) [inline library/cartridge_restore_instance.py:53]: Do you have unit tests for this function?

File: doc/steps.md (hunk 13 above)
Comment (hackallcode) [reply doc/steps.md:590]: I think it's too difficult, and not necessary.

**Review by dokshina — APPROVED** [review]

**Pull request merged**